test/run_portability_test.py

```diff
@@ -0,0 +1,199 @@
+    loaded_dependencies = []
+    dependencies_dict = json.loads(dependencies_json_string)
+
+    for key, value in dependencies_dict.items():
+        try:
+            loaded_dependencies.append({
+                "name": key,
+                "code": open(value).read()
+            })
+        except IOError:
+            errors.append("Count not read dependency {}".format(value))
+
+    if errors:
+        for error in errors:
+            sys.stderr.write(error + "\n")
+        raise TestDefinitionError(errors)
+
+    return {
+        "workflow_descriptor": test_wdl_string,
+        "workflow_params": inputs_json_string,
+        "workflow_dependencies": loaded_dependencies
+        }
+
+def verify_environment_variables():
+    """Check that required environment variables are defined."""
+
+    errors = []
+
+    for key, value in ENV_VARIABLES.items():
+        if not value:
+            errors.append("Environment variable {} is undefined".format(key.upper()))
+    if errors:
+        for error in errors:
+            sys.stderr.write(error + "\n")
+        raise TestEnvironmentError(errors)
+
+def submit_portability_test(test_inputs):
+    """Submit the portability to the service and return the test's id."""
+
+    service_headers = json.loads(ENV_VARIABLES["portability_service_headers"])
+    test_endpoint = ENV_VARIABLES["portability_service_url"] + "/portability_tests"
+    response = requests.post(
+        test_endpoint,
+        headers=service_headers,
+        json=test_inputs)
+
+    print("Portability service response:\n{}".format(
+        json.dumps(json.loads(response.text), indent=4)), flush=True)
+
+    test_id = json.loads(response.text)["test_id"]
+
+    return test_id
+
+def print_test_states(test_states):
+    """Print the states of tests."""
+    print("{:%Y-%m-%dT%H:%M:%SZ}".format(datetime.datetime.utcnow()), flush=True)
+    for test_name, test_state in test_states.items():
+        print("TEST:{} STATE:{}".format(test_name, test_state), flush=True)
+
+def monitor_tests(test_ids):
+    """Check the status of tests until they all either fail or succeed. If any fail, raise
+    TestFailure, and if they all succeed, return successfully.
+    """
+
+    service_headers = json.loads(ENV_VARIABLES["portability_service_headers"])
+
+    terminal_tests = {}
+    while True:
+        time.sleep(120)
+
+        test_states = {}
+
+        for test_name, test_id in test_ids.items():
+
+            if test_name in terminal_tests:
+                test_states[test_name] = terminal_tests[test_name]
+                continue
+
+            status_endpoint = ENV_VARIABLES["portability_service_url"]  + "/portability_tests/" + \
+                test_id + "/status"
+
+            response = requests.get(status_endpoint, headers=service_headers)
+            test_state = json.loads(response.text)["state"]
+            test_states[test_id] = test_state
+
+        print_test_states(test_states)
```

Comment:
Is it possible to log the failures the same way you do for inputs? I'm wondering about a case where you have more than one failing test and want to be explicit about _all_ the failing tests. I see that the print information will contain this but I wonder if the user will notice over the `TestFailure()`.

Reply:
I'm pretty sure that's how it does work. The `TestFailure` mostly just gives the whole script a non-zero exit code if any of the tests fail. That way you get a failure reported via CircleCI.

But, it should run each test to completion and report the status of each one. So if you do see an overall failure, you can look in the CI log and see what test(s) failed. It looks like this:

```
TEST:test/smartseq2_single_sample/pr/ STATE:Succeeded
TEST:test/optimus/pr STATE:Failed
```

Reply:
Ah, got it. Great.